### PR TITLE
Cache cleanup

### PR DESCRIPTION
**Pull request by samdark**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | no
| Breaks BC?    | yes
| Tests pass?   | no
| Fixed issues  | -

### CODE REVIEW

File: src/FileCache.php [hunk 1]
```diff
@@ -135,27 +172,43 @@ protected function setValue($key, $value, $ttl): bool
         $this->gc();
         $cacheFile = $this->getCacheFile($key);
         if ($this->directoryLevel > 0) {
-            @FileHelper::createDirectory(\dirname($cacheFile), $this->dirMode, true);
+            try {
+                @FileHelper::createDirectory(\dirname($cacheFile), $this->dirMode, true);
+            } catch (\yii\exceptions\Exception $error) {
```
Comment (samdark) [inline src/FileCache.php]: Why would it throw an exception there if `@` is used?

File: src/FileCache.php [hunk 2]
```diff
@@ -135,27 +172,43 @@ protected function setValue($key, $value, $ttl): bool
...
+        try {
+            // this implementation is more accurate than file_put_contents
+            // because it does not modify file content, if failed to read
+            if ($fd = fopen($cacheFile, 'cb'))
+            {
+                if (!flock($fd, LOCK_EX)) {
+                    throw new \RuntimeException("Failed to flock '$cacheFile'");
+                } else if(!ftruncate($fd, 0)) {
+                    throw new \RuntimeException("Failed to truncate '$cacheFile");
+                } else if(file_put_contents($cacheFile, $value) !== StringHelper::byteLength($value)) {
```
Comment (samdark) [inline src/FileCache.php]: Does it make sense to use `file_put_contents` if we have opened file descriptor?

File: src/Cache.php [hunk 3]
```diff
@@ -62,12 +61,12 @@
...
  * and [PSR-16 specification](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-16-simple-cache.md).
  */
-class Cache extends Component implements CacheInterface
+class Cache implements CacheInterface
 {
     /**
      * @var \Psr\SimpleCache\CacheInterface actual cache handler.
      */
-    public $_handler;
+    private $_handler;
```
Comment (samdark) [inline src/Cache.php]:
```suggestion
    private $handler;
```

File: src/FileCache.php [hunk 4]
```diff
@@ -135,27 +172,43 @@ protected function setValue($key, $value, $ttl): bool
...
+                    throw new \RuntimeException("Failed to write data to '$cacheFile' totally");
+                }
+                if ($ttl <= 0) {
+                    $ttl = self::NEGATIVE_TTL_REPLACEMENT;
+                }
+                $mtimeInstallationResult = touch($cacheFile, time() + $ttl);
+                // @todo decide, whether to drop file if touched has failed
```
Comment (samdark) [inline src/FileCache.php]: Removing a file seems to be a good idea.

File: src/SimpleCache.php [hunk 5]
```diff
@@ -64,17 +61,25 @@ abstract class SimpleCache extends Component implements CacheInterface
...
+        return new PhpSerializer();
     }
 
     /**
      * @var SerializerInterface the serializer to be used for serializing and unserializing of the cached data.
      *
-     * Default is [[PhpSerializer]], meaning using the default PHP `serialize()` and `unserialize()` functions.
      */
-    public function setSerializer(SerializerInterface $serializer = null)
+    public function setSerializer(SerializerInterface $serializer)
```
Comment (samdark) [inline src/SimpleCache.php]:
```suggestion
    public function setSerializer(SerializerInterface $serializer): void
```

File: src/SimpleCache.php [hunk 6]
```diff
@@ -271,4 +276,32 @@ protected function setValues($values, $ttl): bool
         }
         return $result;
     }
+
+    /**
+     * @param int $defaultTtl
+     */
+    public function setDefaultTtl(int $defaultTtl): void
+    {
+        // @todo validate (possibly everything below 0 is forbidden)
```
Comment (samdark) [inline src/SimpleCache.php]: It is.

File: src/SimpleCache.php [hunk 7]
```diff
@@ -271,4 +276,32 @@ protected function setValues($values, $ttl): bool
...
+        if (!empty($keyPrefix) && !ctype_alnum($keyPrefix)) {
+            throw new \InvalidArgumentException("Given key prefix is not alpha-numeric");
+        }
+        $this->keyPrefix = $keyPrefix;
+    }
+
+    /**
+     * @return int
+     */
+    public function getDefaultTtl(): int
```
Comment (samdark) [inline src/SimpleCache.php]: Does it need to be public?

File: tests/CacheTestCase.php [hunk 8]
```diff
@@ -47,49 +48,91 @@ abstract class CacheTestCase extends TestCase
      */
     public static $microtime;
 
+    /**
+     * List of extensions, that underlying cache requires
+     * If one of extensions is missing - skip test
+     * @var string[]
+     */
+    protected static $required_extensions = [];
```
Comment (samdark) [inline tests/CacheTestCase.php]:
```suggestion
    protected static $requiredExtensions = [];
```

File: src/FileCache.php (hunk 2 above)
Comment (alex-krash) [reply src/FileCache.php]: In order not to write following code in user-space:
```
public function fwriteAll($fp, $str)
{
    if (!mb_orig_strlen($str)) return true;
    $bytes_sent = 0;
    $chunk_size = 65536;
    $write = array($fp);
    $read = $except = null;
    while (stream_select($read, $write, $except, null)) {
        $chunk = substr($str, $bytes_sent, $chunk_size);
        $result = fwrite($fp, $chunk);
        if ($result === false || $result === 0) {
            return false;
        }

        $bytes_sent += $result;
        if ($bytes_sent >= strlen($str)) return true;
        $write = array($fp);
    }

    return true;
}
```

File: src/FileCache.php (hunk 1 above)
Comment (alex-krash) [reply src/FileCache.php]: Adding "@" does not suppress exceptions. It only suppress side-effects of built-in functions. 
I am not sure about "yiisoft/cache" error handling policy (whether error-handler is registered to convert any warning to exceptions). 
For this particular case, I discovered that ::createDirectory can possibly throw an exception, which bubbling is unwanted in this particular case, and decided to wrap it with try/catch.

Comment (samdark) [reply src/FileCache.php]: In which case it may throw an exception? Is it a good idea to ignore this case?

File: src/SimpleCache.php [hunk 9]
```diff
@@ -271,4 +276,32 @@ protected function setValues($values, $ttl): bool
...
+        // @todo validate (possibly everything below 0 is forbidden)
+        $this->defaultTtl = $defaultTtl;
+    }
+
+    /**
+     * @param string $keyPrefix
+     */
+    public function setKeyPrefix(string $keyPrefix): void
+    {
+        if (!empty($keyPrefix) && !ctype_alnum($keyPrefix)) {
```
Comment (samdark) [inline src/SimpleCache.php]: That would forbid names like `cache.name`, `cache_name`, `cache-name`. Is there a technical reason to forbid using any string as a key?

Comment (alex-krash) [reply src/SimpleCache.php]: Let's remove.

File: src/FileCache.php (hunk 1 above)
Comment (alex-krash) [reply src/FileCache.php]: I am not sure, whether we want cache functionality to throw exceptions outside.

**Comment by samdark** [comment]
Merged. Will push soon.

**Pull request closed without merging**